Feature/new media settings for aztec

khaykov opened the pull request:
This PR replaces the old media settings in Aztec with new ones, used in Media Browser.
I also had to fix a couple of things around the media settings and Aztec so this PR turned out a bit bigger than expected.

**Notes:**
Caption and Link support require work on Aztec side, so they are not implemented in this PR.

Editor and Media Browser use similar, but different data models. To reduce branching I convert part of the data to `PostModel`.

I had to change the way we handle image dimensions - editor does not know the real size of an image without loading it, so media settings will only know the real dimensions after loading the image. Until image is loaded the width SeekBar will be greyed out.

Design wise, we are having normal controls with labels alongside control/label combo provided by `TextInputEditText`. Not sure if this gives the best look to editor related controls.

_Empty fields situation looks like this:_
[![https://gyazo.com/a94889122e9fe7728eba2470365204e2](https://i.gyazo.com/a94889122e9fe7728eba2470365204e2.png)](https://gyazo.com/a94889122e9fe7728eba2470365204e2)

_And fields with text look like this:_
[![https://gyazo.com/63b77e061030a0dab6678e6d79ccf093](https://i.gyazo.com/63b77e061030a0dab6678e6d79ccf093.png)](https://gyazo.com/63b77e061030a0dab6678e6d79ccf093)

Hints that turn into labels inside `TextInputEditText` do look a bit weird in this scenario.

My first bigger PR for wpandroid in a while, so hope I haven't messed up anything :)

nbradbury commented:
Great to this this, @khaykov - thanks for tackling it!

As per our Slack discussion, instead of editing the actual pixel width we should show a dropdown similar to iOS, which contains a fixed set of sizes (Thumbnail, Medium, Large, Full Size). If you want to work on that as a separate PR that's fine, but we wouldn't want to merge into develop until that's done (so maybe a master/feature branch)?

nbradbury commented:
I think the excess white space here could be fixed by hiding the containing `TextInputLayout` instead of the `TextInputEditText`.

![screenshot_1510064449](https://user-images.githubusercontent.com/3903757/32501834-16a5653a-c3a7-11e7-93e2-d8a04af58585.png)

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/media/MediaSettingsActivity.java:
```diff
@@ -419,6 +504,10 @@ private void showProgress(boolean show) {
+    }
+
+    private void finishActivity() {
         // call finish() rather than super.onBackPressed() to enable skipping shared element transition
```
I think this comment should be in `onBackPressed()` above.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/media/MediaSettingsActivity.java:
```diff
@@ -264,12 +316,22 @@ public void onClick(View v) {
     private boolean loadMediaId(int mediaId) {
-        MediaModel media = mMediaStore.getMediaWithLocalId(mediaId);
+        MediaModel media;
+        if (isMediaFromEditor()) {
```
This method signature seems off to me, since the passed `mediaId` is ignored when the media came from the editor. Any way we can change this without over-complicating things?

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/media/MediaSettingsActivity.java:
```diff
@@ -313,6 +375,19 @@ private boolean loadMediaId(int mediaId) {
+        mediaModel.setTitle(mEditorImageMetaData.getTitle());
+        mediaModel.setCaption(mEditorImageMetaData.getCaption());
+        mediaModel.setAlt(mEditorImageMetaData.getAlt());
+        mediaModel.setFileName(mEditorImageMetaData.getSrc().substring(mEditorImageMetaData.getSrc().lastIndexOf("/") + 1));
```
Should we null check `getSrc()` here before calling `substring()` on it?

khaykov replied:
Technically, the `src` should never be null, because we are making json internally ([here])(https://github.com/wordpress-mobile/WordPress-Android/blob/develop/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/MetadataUtils.java#L22) and always use empty strings as the default value. But you are right, and it's probably better to be safe, in case it's going to get changed at some point.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/media/MediaSettingsActivity.java:
```diff
@@ -745,32 +948,76 @@ public void onReceive(Context context, Intent intent) {
-            media.setAlt(thisAltText);
-            media.setDescription(thisDescription);
-            mDispatcher.dispatch(MediaActionBuilder.newPushMediaAction(new MediaStore.MediaPayload(mSite, media)));
+        if (!isMediaFromEditor()) {
```
Just a heads-up that at some point we'll want to save changes to the media store even when shown from the editor. So, for example, if the user inserts an image from the WP media library and then edits the alt text, we'll want to save the alt text back to the store. That's outside the scope of this PR, though.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/media/MediaSettingsActivity.java:
```diff
@@ -95,16 +104,24 @@
+    private Spinner mAlignmentSpinnerView;
     private FloatingActionButton mFabView;
 
+    private AlertDialog mDeleteImageConfirmationDialog;
```
Would it simplify things if instead of two `AlertDialogs` we just had one whose text & action depend on whether the media was shown from the editor?

nbradbury commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/EditorImageMetaData.java:
```diff
@@ -0,0 +1,336 @@
+    private int mLocalId;
+
+    public String getAlign() {
+        if (mAlign.startsWith(ATTR_ALIGN)) {
```
Do we need to null check `mAlign` here?

nbradbury commented:
Can we get rid of the left padding on the spinner item, so it aligns with the labels? This will most likely require using a custom view for the spinner adapter.

![screenshot_1510075058](https://user-images.githubusercontent.com/3903757/32508480-77600784-c3b8-11e7-86c5-453bc7a850df.png)

nbradbury commented:
@khaykov I'm done with my first round, ping me when you're ready for round two. I'm really glad you came on board right when we needed someone with editor experience to tackle this - it would've taken me a lot longer to figure it out.

khaykov commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/EditorImageMetaData.java:
```diff
@@ -0,0 +1,336 @@
+
+    public int getHeightInt() {
+        try {
+            return Integer.parseInt(mHeight);
```
Thanks! Didn't know about it.

khaykov commented:
@nbradbury Thanks for the review! I'm ready for another round :)

nbradbury commented:
Minor nit: it looks like we're inserting empty attribute values for title, alt text, etc., when the user hasn't provided them. It would be better if we simply left these attributes out when empty.

![screenshot_1510409935](https://user-images.githubusercontent.com/3903757/32729150-cc757280-c850-11e7-9256-912b35f31b4b.png)

daniloercoli commented:
![device-2017-11-13-162332](https://user-images.githubusercontent.com/518232/32733596-005195f4-c890-11e7-95cf-99758782bb85.png)

This is more a question for @iamthomasbishop(@khaykov and @nbradbury feel free to reply if you've already discussed this, and sorry If i'm adding details without having follows the design process): Wondering if we can switch the "X" (back button) on the toolbar with a left arrow.... I was testing this PR for the first time, and was a bit lost, since I had modified the title and the alt text and was not sure which one of the 2 button had to press to go back to the editor without loosing all of the changes. The "X" seemed like I was discard all the changes I made to the settings.

nbradbury commented:
I'm okay with changing the X to a back arrow - no strong feelings about it. But for me the confusing part is that "remove" icon because it's not immediately obvious what it does.

daniloercoli commented:
With the current implementation long pressing the "remove" icon shows a toast with just "Remove" text in it. Should we also add a better description there, regardless on the decision on the icon?

khaykov commented:
@nbradbury I added a check for empty attributes and updated the Remove button hint.
As we discussed, I will address the issue @daniloercoli brought up on Aztec side.

nbradbury commented:
Looks good, great work on this :shipit:

iamthomasbishop commented:
@daniloercoli we chose a flow like this on iOS and should probably do similar on android:

1. User taps on `trash icon`
2. Sees options for `remove from this post` or `delete from media library`
3. If they choose to delete from their library altogether, they should have a final prompt to confirm they'd like to delete it and note that they can't undo this action. If they choose to just remove from the post, we only show confirmation. (Maybe we should show confirmation in both cases - or only in the delete scenario?), perhaps in a snackbar format - thoughts?)

It looks something like this on iOS:

<img width="1486" alt="screenshot 2017-11-15 13 46 21" src="https://user-images.githubusercontent.com/1200432/32856868-b56873b4-ca0b-11e7-9d14-b5dc171f8622.png">

We could probably show a simple menu up top or a bottom sheet when the user taps the trash icon, unless there are other options that are more appropriate. 

Note that the user should be able to both remove from the post _and_ delete completely from library. Because of this, the `trash` icon is what we should use in the top right corner.

daniloercoli commented:
Thanks for your reply @iamthomasbishop, the new flow seems very solid! 💯
